velox/functions/lib/aggregates/AverageAggregateBase.h

```diff
@@ -0,0 +1,417 @@
+    DecodedVector baseCountDecoded{*baseRowVector->childAt(1), *baseRows};
+
+    if (decodedPartial_.isConstantMapping()) {
+      if (!decodedPartial_.isNullAt(0)) {
+        auto decodedIndex = decodedPartial_.index(0);
+        if constexpr (checkNullFields) {
+          VELOX_USER_CHECK(
+              !baseSumDecoded.isNullAt(decodedIndex) &&
+              !baseCountDecoded.isNullAt(decodedIndex));
+        }
+        const auto numRows = rows.countSelected();
+        auto totalCount =
+            baseCountDecoded.template valueAt<int64_t>(decodedIndex) * numRows;
+        auto totalSum =
+            baseSumDecoded.template valueAt<TAccumulator>(decodedIndex) *
+            numRows;
+        updateNonNullValue(group, totalCount, totalSum);
+      }
+    } else if (decodedPartial_.mayHaveNulls()) {
+      rows.applyToSelected([&](vector_size_t i) {
+        if (!decodedPartial_.isNullAt(i)) {
+          auto decodedIndex = decodedPartial_.index(i);
+          if constexpr (checkNullFields) {
+            VELOX_USER_CHECK(
+                !baseSumDecoded.isNullAt(decodedIndex) &&
+                !baseCountDecoded.isNullAt(decodedIndex));
+          }
+          updateNonNullValue(
+              group,
+              baseCountDecoded.template valueAt<int64_t>(decodedIndex),
+              baseSumDecoded.template valueAt<TAccumulator>(decodedIndex));
+        }
+      });
+    } else {
+      TAccumulator totalSum(0);
+      int64_t totalCount = 0;
+      rows.applyToSelected([&](vector_size_t i) {
+        auto decodedIndex = decodedPartial_.index(i);
+        if constexpr (checkNullFields) {
+          VELOX_USER_CHECK(
+              !baseSumDecoded.isNullAt(decodedIndex) &&
+              !baseCountDecoded.isNullAt(decodedIndex));
+        }
+        totalCount += baseCountDecoded.template valueAt<int64_t>(decodedIndex);
+        totalSum += baseSumDecoded.template valueAt<TAccumulator>(decodedIndex);
+      });
+      updateNonNullValue(group, totalCount, totalSum);
+    }
+  }
+
+  DecodedVector decodedRaw_;
+  DecodedVector decodedPartial_;
+};
+
+template <typename TUnscaledType>
+class DecimalAverageAggregateBase
+    : public velox::aggregate::DecimalAggregate<TUnscaledType> {
+ public:
+  explicit DecimalAverageAggregateBase(TypePtr resultType)
+      : velox::aggregate::DecimalAggregate<TUnscaledType>(resultType) {}
+
+  virtual TUnscaledType computeFinalValue(
+      velox::aggregate::LongDecimalWithOverflowState* accumulator) final {
+    // Handles round-up of fraction results.
+    int128_t average{0};
+    DecimalUtil::computeAverage(
+        average, accumulator->sum, accumulator->count, accumulator->overflow);
+    return TUnscaledType(average);
+  }
+};
+
+/// @brief Checks the input type for final aggregation of average.
+/// The input type must be (sum:double/long decimal, count:bigint) struct.
+/// @param type input type for final aggregation of average.
+void checkAvgFinalType(const TypePtr& type);
```

Comment:
I just realized that 'type' here is not final type, but rather intermediate type. Perhaps, rename to checkAvgIntermediateType.